### PR TITLE
Logging: Unify usage of log levels

### PR DESCRIPTION
**Pull request by EntranceJew**
if we expect people to fix them then we should help them find them

### CODE REVIEW

File: gamemodes/terrortown/gamemode/server/sv_addonchecker.lua [hunk 1]
```diff
@@ -651,16 +651,16 @@ function addonChecker.Check()
 		local detectedAddon = addonChecker.curatedList[tostring(addon.wsid)]
 		if not detectedAddon then continue end
 
-		ErrorNoHalt(((detectedAddon.type == ADDON_OUTDATED) and "Outdated add-on detected: " or "Incompatible add-on detected: ") .. addon.title .. "\n")
+		ErrorNoHaltWithStack(((detectedAddon.type == ADDON_OUTDATED) and "Outdated add-on detected: " or "Incompatible add-on detected: ") .. addon.title .. "\n")
 
 		if detectedAddon.reason then
-			ErrorNoHalt("Reason: " .. detectedAddon.reason .. "\n")
+			ErrorNoHaltWithStack("Reason: " .. detectedAddon.reason .. "\n")
```
Comment (TimGoll) [inline gamemodes/terrortown/gamemode/server/sv_addonchecker.lua:657]: Im this file it makes no sense

File: gamemodes/terrortown/gamemode/server/sv_karma.lua [hunk 2]
```diff
@@ -635,7 +635,7 @@ function KARMA.Remember(ply)
 	end
 
 	if ply:SteamID64() == nil then
-		print("[TTT2] ERROR: Player has no steamID64")
+		ErrorNoHaltWithStack("[TTT2] ERROR: Player has no steamID64:" .. ply)
```
Comment (TimGoll) [inline gamemodes/terrortown/gamemode/server/sv_karma.lua:638]: Does this work without a tostring

File: gamemodes/terrortown/gamemode/server/sv_main.lua [hunk 3]
```diff
@@ -356,13 +356,13 @@ function GM:Initialize()
 	WaitForPlayers()
 
 	if cvars.Number("sv_alltalk", 0) > 0 then
-		ErrorNoHalt("TTT2 WARNING: sv_alltalk is enabled. Dead players will be able to talk to living players. TTT2 will now attempt to set sv_alltalk 0.\n")
+		ErrorNoHaltWithStack("TTT2 WARNING: sv_alltalk is enabled. Dead players will be able to talk to living players. TTT2 will now attempt to set sv_alltalk 0.\n")
```
Comment (TimGoll) [inline gamemodes/terrortown/gamemode/server/sv_main.lua:359]: I think a stack isn't useful here

File: gamemodes/terrortown/gamemode/server/sv_main.lua [hunk 4]
```diff
@@ -356,13 +356,13 @@ function GM:Initialize()
...
 
 		RunConsoleCommand("sv_alltalk", "0")
 	end
 
 	if not IsMounted("cstrike") then
-		ErrorNoHalt("TTT2 WARNING: CS:S does not appear to be mounted by GMod. Things may break in strange ways. Server admin? Check the TTT readme for help.\n")
+		ErrorNoHaltWithStack("TTT2 WARNING: CS:S does not appear to be mounted by GMod. Things may break in strange ways. Server admin? Check the TTT readme for help.\n")
```
Comment (TimGoll) [inline gamemodes/terrortown/gamemode/server/sv_main.lua:365]: Same here

File: gamemodes/terrortown/gamemode/shared/sh_equip_items.lua [hunk 5]
```diff
@@ -726,7 +726,7 @@ function GenerateNewEquipmentID()
 
 			if v.oldId ~= val or not v.id then continue end
 
-			print("[TTT2][WARNING] TTT2 doesn't support old items completely since they are limited to an amount of 16. If the item '" .. v.id .. "' with id '" .. val .. "' doesn't work as intended, modify the old item and use the new items system instead.")
+			ErrorNoHaltWithStack("[TTT2][WARNING] TTT2 doesn't support old items completely since they are limited to an amount of 16. If the item '" .. v.id .. "' with id '" .. val .. "' doesn't work as intended, modify the old item and use the new items system instead.")
```
Comment (TimGoll) [inline gamemodes/terrortown/gamemode/shared/sh_equip_items.lua:729]: Stack isn't helpful here I think

File: lua/ttt2/libraries/items.lua [hunk 6]
```diff
@@ -172,7 +172,7 @@ function items.Get(name, retTbl)
 		local base = items.Get(retval.Base)
 
 		if not base then
-			Msg("ERROR: Trying to derive item " .. tostring(name) .. " from non existant item " .. tostring(retval.Base) .. "!\n")
+			ErrorNoHaltWithStack("ERROR: Trying to derive item " .. tostring(name) .. " from non existant item " .. tostring(retval.Base) .. "!\n")
```
Comment (TimGoll) [inline lua/ttt2/libraries/items.lua:175]: As this is also only internally called, I'm not sure a stack is any helpful here

File: lua/ttt2/libraries/plyspawn.lua [hunk 7]
```diff
@@ -203,7 +203,7 @@ function plyspawn.GetRandomSafePlayerSpawnPoint(ply)
 
 		if not riggedSpawnPoint then continue end
 
-		ErrorNoHalt("TTT2 WARNING: Map has too few spawn points, using a riggedSpawnPoints spawn for " .. tostring(ply) .. "\n")
+		ErrorNoHaltWithStack("TTT2 WARNING: Map has too few spawn points, using a riggedSpawnPoints spawn for " .. tostring(ply) .. "\n")
```
Comment (TimGoll) [inline lua/ttt2/libraries/plyspawn.lua:206]: This is on the map maker and happens quite a lot. Please remove that here, there is nothing the dev can change about that

**Review by TimGoll — CHANGES_REQUESTED** [review]
This is already quite nice, but there are some places where this shouldn't be changed. I might have missed some, maybe you should check them yourself again and think if this adds any benefit.

Also I'm not sure if we should add the trace to the language stuff as these old addons will likely never be updated and my console is already flooded with depreciation warnings.

**Comment by EntranceJew** [comment]
FWIW my point of reference for severity is as follows:
- **ErrorNoHaltWithStack** things that should be fixed, or would otherwise get an addon put in the addonchecker.
- **Dev(1, ...)** anything that qualifies as an indication of program flow / features being debugged
- **Dev(2, ...)** extraneous debugging noise
- **Dev(N, ...)** even noisier noise
- **print** should only happen as a direct consequence of debug commands 

banished:
- **MsgN**does not convey a priority
- **ErrorNoHalt** functionally the same as MsgN, but with an error connotation
- **Error** identical to ErrorNoHalt, easy to mistake for `error`, therefore, same treatment
- **error** if we're going to error, we may as well promote a stack via **ErrorNoHaltWithStack** -- unless we need to redirect blame via `errorlevel`

**Comment by TimGoll** [comment]
I absolutely agree with you. My raised points still hold true because we don't want to spam the console when internal code raises these flags as the developers have no chance of fixing it

File: lua/ttt2/libraries/items.lua (hunk 6 above)
Comment (EntranceJew) [reply lua/ttt2/libraries/items.lua:175]: If it happens at all, it should be fixed, right? Some indication of how someone got there is important. I definitely utilize this inside PotOfGreedier.

**Comment by TimGoll** [comment]
But the ones I marked are unfixable, like this one: 
image

this could only be fixed by the map dev, but this isn't going to happen

**Comment by EntranceJew** [comment]
> this could only be fixed by the map dev, but this isn't going to happen. Also a trace is not useful at all in this case

i hear you, i just can only respond to things to keep track of what i changed and didn't change, might contest individual situations :U

File: gamemodes/terrortown/gamemode/server/sv_addonchecker.lua
Comment (Histalek) [inline gamemodes/terrortown/gamemode/server/sv_addonchecker.lua:1]: Please revert the changes in this file. We want people to always see the Output of the addon checker without any extra steps involved.

**Review by Histalek — CHANGES_REQUESTED** [review]
Anything besides the addonchecker changes looks good :)

File: gamemodes/terrortown/entities/entities/ttt_credit_adjust.lua [hunk 8]
```diff
@@ -20,7 +20,7 @@ function ENT:KeyValue(key, value)
 		self.Credits = tonumber(value) or 0
 
 		if not tonumber(value) then
-			ErrorNoHalt(tostring(self) .. " has bad 'credits' setting.\n")
+			ErrorNoHaltWithStack(tostring(self) .. " has bad 'credits' setting.\n")
```
Comment (TimGoll) [inline gamemodes/terrortown/entities/entities/ttt_credit_adjust.lua:23]: Could you explain here the helpfulness of this? Because this is also called from a map entity, meaning that the stacktrace will probably not point to any error that actually exists.

I'm not entirely sure though, that's just my impression. Feel free to say I'm wrong here

File: gamemodes/terrortown/entities/entities/ttt_game_text.lua [hunk 9]
```diff
@@ -39,7 +39,7 @@ function ENT:KeyValue(key, value)
 		self.Receiver = tonumber(value)
 
 		if not self.Receiver or self.Receiver < 0 or self.Receiver > 5 then
-			ErrorNoHalt("ERROR: ttt_game_text has invalid inputReceiver value\n")
+			ErrorNoHaltWithStack("ERROR: ttt_game_text has invalid inputReceiver value\n")
```
Comment (TimGoll) [inline gamemodes/terrortown/entities/entities/ttt_game_text.lua:42]: same here for these two. I'm not saying this is wrong, I just wonder if it adds any meaningful information here?

File: gamemodes/terrortown/entities/entities/ttt_logic_role.lua [hunk 10]
```diff
@@ -29,7 +29,7 @@ function ENT:KeyValue(key, value)
 		self.checkingRole = tonumber(value)
 
 		if not self.checkingRole then
-			ErrorNoHalt("ttt_logic_role: bad value for Role key, not a number\n")
+			ErrorNoHaltWithStack("ttt_logic_role: bad value for Role key, not a number\n")
```
Comment (TimGoll) [inline gamemodes/terrortown/entities/entities/ttt_logic_role.lua:32]: and here as well, this is also a map entity that is never triggered inside lua. This is an error the mapmaker has to take care of, hence I think the stack is useless

File: gamemodes/terrortown/entities/entities/ttt_map_settings.lua [hunk 11]
```diff
@@ -66,11 +66,11 @@ function ENT:AcceptInput(name, activator, caller, data)
 		local mdlname = tostring(data)
 
 		if not mdlname then
-			ErrorNoHalt("ttt_map_settings: Invalid parameter to SetPlayerModels input!\n")
+			ErrorNoHaltWithStack("ttt_map_settings: Invalid parameter to SetPlayerModels input!\n")
```
Comment (TimGoll) [inline gamemodes/terrortown/entities/entities/ttt_map_settings.lua:69]: here as well, this is also a map ent and a few lines down below as well

File: gamemodes/terrortown/entities/entities/ttt_weapon_check.lua [hunk 12]
```diff
@@ -102,7 +102,7 @@ function ENT:TestWeapons(weptype)
 	local check = self:GetWeaponChecker(weptype)
 
 	if check == nil then
-		ErrorNoHalt("ttt_weapon_check: invalid parameter\n")
+		ErrorNoHaltWithStack("ttt_weapon_check: invalid parameter\n")
```
Comment (TimGoll) [inline gamemodes/terrortown/entities/entities/ttt_weapon_check.lua:105]: here as well, these two checks are map entities

File: gamemodes/terrortown/gamemode/client/cl_lang.lua [hunk 13]
```diff
@@ -69,7 +69,7 @@ function LANG.AddToLanguage(langName, stringName, stringText)
 	langName = LANG.GetNameFromAlias(langName)
 
 	if not LANG.IsLanguage(langName) then
-		ErrorNoHalt(Format("Failed to add '%s' to language '%s': language does not exist.\n", tostring(stringName), tostring(langName)))
+		ErrorNoHaltWithStack(Format("Failed to add '%s' to language '%s': language does not exist.\n", tostring(stringName), tostring(langName)))
```
Comment (TimGoll) [inline gamemodes/terrortown/gamemode/client/cl_lang.lua:72]: as I said in my last review, my console is already spammed with lang deprecation warning. If we add t more lines to each warning I think the console becomes unusable, don't you think?

File: gamemodes/terrortown/gamemode/client/cl_radio.lua [hunk 14]
```diff
@@ -273,21 +273,21 @@ end
...
+		ErrorNoHaltWithStack("ttt_radio failed, invalid player")
 
 		return
 	end
 
 	---
 	-- @realm client
 	if hook.Run("TTT2ClientRadioCommand", cmd) then
-		print("ttt_radio, execution prevented by a hook")
+		ErrorNoHaltWithStack("ttt_radio, execution prevented by a hook")
```
Comment (TimGoll) [inline gamemodes/terrortown/gamemode/client/cl_radio.lua:284]: no, this is bad. If a hook prevents the radio, it shouldn't throw an error. But tbh, this print should probably be removed or moved to a MStack message

File: gamemodes/terrortown/gamemode/server/sv_addonchecker.lua
Comment (TimGoll) [reply gamemodes/terrortown/gamemode/server/sv_addonchecker.lua:1]: agreed, the addonchecker should always be visible to everyone

**Review by TimGoll — COMMENTED** [review]
This is now mostly good, thanks! There are a few questions and one major issue here

File: gamemodes/terrortown/gamemode/client/cl_lang.lua (hunk 13 above)
Comment (TimGoll) [reply gamemodes/terrortown/gamemode/client/cl_lang.lua:72]: well this comment is tied to the wrong line, sorry about that. This specific change here is good.

But in the same file is a deprecation warning. I'm talking about the one in line 223, the rest seems fine

File: gamemodes/terrortown/gamemode/client/cl_radio.lua (hunk 14 above)
Comment (Histalek) [reply gamemodes/terrortown/gamemode/client/cl_radio.lua:284]: If this should be moved to a MStack message this would also need a lang string.

i'd say reverting this for now matches the scope of this PR.

Comment (TimGoll) [reply gamemodes/terrortown/gamemode/client/cl_radio.lua:284]: or put it as a DEV(1 for now

File: gamemodes/terrortown/gamemode/client/cl_lang.lua (hunk 13 above)
Comment (EntranceJew) [reply gamemodes/terrortown/gamemode/client/cl_lang.lua:72]: the problem is that this error does not indicate the oirigin of the problem. even if i wanted to fix this, i would need to unpack every addon i have in order to locate the problematic language file, and it's so far removed from the source file directory that it cannot be associated _without_ a stack trace

File: gamemodes/terrortown/gamemode/server/sv_addonchecker.lua
Comment (EntranceJew) [reply gamemodes/terrortown/gamemode/server/sv_addonchecker.lua:1]: > This looks mostly sane to me, i would change 2 things though.
> 
> * `print()` should not be in our codebase

wat do

Comment (Histalek) [reply gamemodes/terrortown/gamemode/server/sv_addonchecker.lua:1]: > > This looks mostly sane to me, i would change 2 things though.
> > 
> > * `print()` should not be in our codebase
> 
> wat do

**should** :D 
besides that, this is a special case as it is more akin to program output and not a logged error/problem. `print` therefore is fine.

**Review by Histalek — APPROVED** [review]

File: gamemodes/terrortown/gamemode/client/cl_lang.lua (hunk 13 above)
Comment (TimGoll) [reply gamemodes/terrortown/gamemode/client/cl_lang.lua:72]: I absolutely understand that. But 99% of the people only want to play TTT2 and will be pretty confused when there are errors on every mapchange that they can't fix

Comment (TimGoll) [reply gamemodes/terrortown/gamemode/client/cl_lang.lua:72]: I reverted this, @Histalek and I talked about a better solution in our debug library that can be done in a coming PR so that we can merge this one!

**Review by TimGoll — APPROVED** [review]
